tools/astarte_e2e/lib/astarte_e2e/client.ex

```diff
@@ -0,0 +1,575 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule AstarteE2E.Client do
+  alias Phoenix.Channels.GenSocketClient
+  alias Phoenix.Channels.GenSocketClient.Transport.WebSocketClient
+  alias AstarteE2E.{Utils, Config}
+
+  require Logger
+
+  @connection_backoff_ms 10_000
+  @connection_attempts 10
+
+  # API
+
+  @doc "Starts the client process."
+  @spec start_link(Config.client_options()) :: GenServer.on_start()
+  def start_link(opts) do
+    transport = WebSocketClient
+    url = Keyword.fetch!(opts, :url)
+    jwt = Keyword.fetch!(opts, :jwt)
+    realm = Keyword.fetch!(opts, :realm)
+    device_id = Keyword.fetch!(opts, :device_id)
+    check_repetitions = Keyword.fetch!(opts, :check_repetitions)
+
+    verify_option =
+      if Keyword.get(opts, :ignore_ssl_errors, false) do
+        :verify_none
+      else
+        :verify_peer
+      end
+
+    remote_device = [
+      url: url,
+      realm: realm,
+      jwt: jwt,
+      device_id: device_id,
+      check_repetitions: check_repetitions
+    ]
+
+    with {:ok, pid} <-
+           GenSocketClient.start_link(
+             __MODULE__,
+             transport,
+             remote_device,
+             [transport_opts: [ssl_verify: verify_option]],
+             name: via_tuple(realm, device_id)
+           ) do
+      :telemetry.execute(
+        [:astarte_end_to_end, :astarte_platform],
+        %{status: 0}
+      )
+
+      Logger.info("Started process with pid #{inspect(pid)}.", tag: "client_started")
+
+      {:ok, pid}
+    end
+  end
+
+  def child_spec(opts) do
+    %{
+      id: __MODULE__,
+      start: {__MODULE__, :start_link, [opts]},
+      type: :worker,
+      restart: :transient,
+      shutdown: 500
+    }
+  end
+
+  @spec verify_device_payload(String.t(), String.t(), String.t(), String.t(), any(), integer()) ::
+          :ok
+          | {:error, :not_connected | :timeout}
+  def verify_device_payload(realm, device_id, interface_name, path, value, timestamp) do
+    GenSocketClient.call(
+      via_tuple(realm, device_id),
+      {:verify_payload, interface_name, path, value, timestamp},
+      :infinity
+    )
+  end
+
+  def wait_for_connection(client_pid) do
+    GenSocketClient.call(client_pid, :wait_for_connection, :infinity)
+  end
+
+  @spec get_pid(String.t(), String.t()) :: pid() | nil
+  def get_pid(realm, device_id) do
+    case Registry.lookup(Registry.AstarteE2E, {:client, realm, device_id}) do
+      [{pid, _}] ->
+        pid
+
+      _ ->
+        Logger.warn("Unregistered client process.", tag: "unregistered_process")
+
+        nil
+    end
+  end
```

Comment:
Since you already have the infrastructure for `via_tuple`, be consistent with the API and do
```elixir
def wait_for_connection(realm, device_id) do
  via_tuple(realm, device_id)
  |> GenSocketClient.call(:wait_for_connection, :infinity)
end
```
so you don't event need `get_pid`